SPEC.md

```diff
@@ -0,0 +1,225 @@
+
+<p align="center">
+  <br>
+  <img src="images/apollo-fetcher.png" alt="Apollo Fetcher"/>
+</p>
+
+Fetchers are designed for current Apollo Client use-cases and to provide room for extensions. In addition to supporting Apollo Client, Apollo Fetchers are designed to be used independently as a very simple GraphQL client. These are the use cases that fetchers should support:
+
+* Do a simple GraphQL query with it that resolves only once
+* Do a simple GraphQL mutation that resolves only once
+* Do a subscription
+* Do a GraphQL query that contains @defer, @stream or @live
+* Send extra metadata along with queries
+* Support middleware to modify headers
+* Support middleware to modify query, variables, operationName and metadata
+* Isomorphic: It works in the browser and on the server
+* Support afterware to modify response
+* Support HTTP  transport, websocket transport, hybrid transport
+* Support alternative transports
+* Support per request retries based on network status or response
+* Support HTTP transport batching
+* Support polling
+* Support batching
+* Request deduplication of on-the-wire requests
+* Support splitting: eg. send queries and mutations to different servers
+
+Since this list focuses on Apollo Client's needs, please open a Pull Request adding to this list as you think of your desired use cases.
+
+In order to support all these use cases with a simple fetcher design, we want to use a modular architecture that allows fetchers to delegate to other fetchers.
+
+## Example
+
+This repository will contain a family of canonical fetchers that provide example extensions. The goal of these canonical fetchers is to be self-contained without changing the context of each operation.
+
+The current sample is http-fetcher, which sends an operation over http to a server at a specific URI. The uri and fetch function are passed to the constructor. This Http-fetcher can be used as a stand-alone GraphQL client:
+
+<p align="center">
+  <br>
+  <img src="images/http-fetcher.png" alt="Apollo Fetcher"/>
+</p>
+
+```js
+const httpFetcher = new HTTPFetcher({ uri: 'http://api.githunt.com/graphql' });
+
+const responseObservable = httpFetcher.request({
+  query,
+  variables,
+  operationName,
+  context,
+});
+
+responseObservable.subscribe({
+  next(data){ console.log('received data', data); },
+  error(error){ console.error(error); },
+  complete(){ console.error('request complete'); },
+});
+```
+
+A fetcher can delegate to other fetchers. For example a polling fetcher can send a query operation to an http fetcher on a specified interval. Each request is routed through that http fetcher and returned through the observable returned by the polling fetcher. Using this type of delegation, fetchers can form stacks that each add their own functionality.
```

Comment:
I feel like I prefer the concept of "composing" rather than "delegating". To me, delegating means that a polling fetcher might use an http fetcher internally, but composing means I connect the polling fetcher and http fetcher myself.